### PR TITLE
Add definition for dicer

### PR DESCRIPTION
**Pull request by rgoupil**
- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [X] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [X] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [X] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [X] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

___

I took the liberty to turn off the tslint rule "unified-signatures" because it would force me to change the signatures of the event definitions from this:

    addListener(event: "finish", listener: () => void): this;
    addListener(event: "close", listener: () => void): this;
    addListener(event: "drain", listener: () => void): this;

to this:

    addListener(event: "finish"|"close"|"drain", listener: () => void): this;

This make the definitions harder to read and to find in auto completion (at least on VSCode).

Finally, I was wondering if I should add a jsdoc for each `on`, `once`, etc. of each event but chose not to because:
* It is harder to read
* It duplicate information a lot
* The definitions for stream.Readable or stream.Writable don't do it either

Sadly, this mean that the events are only documented at the class jsdoc and not at each event.

Please let me know f you think I made the wrong choices.

### CODE REVIEW

**Review by plantain-00 — CHANGES_REQUESTED** [review]

**Review by rgoupil — COMMENTED** [review]
It looks bad, but `tslint:disable-next-line` looked even worst 😢

**Comment by GoupilRobin** [comment]
> typescript-bot added the Revision needed label 30 minutes ago

Is there something else I should fix ?
I don't see any comment 😕

**Review by plantain-00 — APPROVED** [review]

File: types/dicer/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+ * A very fast streaming multipart parser for node.js.
+ * Dicer is a WritableStream
+ *
+ * Dicer (special) events:
+ * - on('finish', ()) - Emitted when all parts have been parsed and the Dicer instance has been ended.
+ * - on('part', (stream: PartStream)) - Emitted when a new part has been found.
+ * - on('preamble', (stream: PartStream)) - Emitted for preamble if you should happen to need it (can usually be ignored).
+ * - on('trailer', (data: Buffer)) - Emitted when trailing data was found after the terminating boundary (as with the preamble, this can usually be ignored too).
+ */
+export default class Dicer extends stream.Writable {
```
Comment (rbuckton) [inline types/dicer/index.d.ts]: The actual export is `module.exports = Dicer`. The equivalent declaration in TypeScript would be `export = Dicer`.

**Review by rbuckton — CHANGES_REQUESTED** [review]

File: types/dicer/index.d.ts (hunk 1 above)
Comment (GoupilRobin) [reply types/dicer/index.d.ts]: AFAIK, using the `export =` notation will force to use `import Dicer = require('dicer')`, while `export default` is much more flexible as it allows the above syntax and other imports syntax, which is nice for people with a preference for the import syntax they use.

I'm fine nesting the interfaces in the `Dicer` class (`DicerConfig` would become `Dicer.Config` and `PartStream` would become `Dicer.PartStream`), but if there is no functionnal difference between the two forms, I would argue that the more flexibility, the better.

**Comment by GoupilRobin** [comment]
@rbuckton thank you for answering my question prompty...

**Comment by GoupilRobin** [comment]
Also @RyanCavanaugh can we remove the "Abandoned" tag, I posted a question on this PR and was simply waiting for an answer from @rbuckton, who didn't provide it in a timely fashion.
Thanks

**Comment by GoupilRobin** [comment]
Thank you @RyanCavanaugh and @rbuckton for not anwering any of my requests.

**Pull request closed without merging**